src/coreclr/jit/lowerxarch.cpp

```diff
@@ -6335,38 +6335,49 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
                 {
                     if ((intrinsicId >= NI_FMA_MultiplyAdd) && (intrinsicId <= NI_FMA_MultiplySubtractNegatedScalar))
                     {
-                        bool supportsRegOptional = false;
+                        bool     supportsOp1RegOptional = false;
+                        bool     supportsOp2RegOptional = false;
+                        bool     supportsOp3RegOptional = false;
+                        unsigned resultOpNum            = 0;
+                        LIR::Use use;
 
-                        if (IsContainableHWIntrinsicOp(node, op3, &supportsRegOptional))
+                        if (BlockRange().TryGetUse(node, &use))
                         {
-                            // 213 form: op1 = (op2 * op1) + [op3]
-                            MakeSrcContained(node, op3);
+                            resultOpNum = node->GetResultOpNumForFMA(use.User(), op1, op2, op3);
                         }
-                        else if (IsContainableHWIntrinsicOp(node, op2, &supportsRegOptional))
+
+                        // Prioritize Containable op. Check if any one of the op is containable first.
+                        // Set op regOptional only if none of them is containable.
+
+                        if (resultOpNum != 1 && IsContainableHWIntrinsicOp(node, op1, &supportsOp1RegOptional) &&
+                            !HWIntrinsicInfo::CopiesUpperBits(intrinsicId))
                         {
-                            // 132 form: op1 = (op1 * op3) + [op2]
+                            // result = ([op1] * op2) + op3
+                            MakeSrcContained(node, op1);
+                        }
+                        else if (resultOpNum != 2 && IsContainableHWIntrinsicOp(node, op2, &supportsOp2RegOptional))
+                        {
+                            // result = (op1 * [op2]) + op3
                             MakeSrcContained(node, op2);
                         }
-                        else if (IsContainableHWIntrinsicOp(node, op1, &supportsRegOptional))
+                        else if (resultOpNum != 3 && IsContainableHWIntrinsicOp(node, op3, &supportsOp3RegOptional))
                         {
-                            // Intrinsics with CopyUpperBits semantics cannot have op1 be contained
-
-                            if (!HWIntrinsicInfo::CopiesUpperBits(intrinsicId))
-                            {
-                                // 231 form: op3 = (op2 * op3) + [op1]
-                                MakeSrcContained(node, op1);
-                            }
+                            // result = (op1 * op2) + [op3]
+                            MakeSrcContained(node, op3);
+                        }
+                        else if (resultOpNum != 1 && !HWIntrinsicInfo::CopiesUpperBits(intrinsicId))
+                        {
+                            assert(supportsOp1RegOptional);
+                            op1->SetRegOptional();
+                        }
```

Comment:
This part is written in this way to prioritize containable op. In the case like ```resultOpNum==2```, we want to check if ```op3``` is containable before we decide to set ```op1``` is regOptional. Please let me know if this is a reasonable consideration.

Reply:
Ah right, because if `resultOpNum == 1` then either `op2` or `op3` is containable.

It would be nice if this could be simplified somehow, but its not immediately obvious what that would be.